vllm/v1/core/sched/scheduler.py

```diff
@@ -653,11 +653,33 @@ def update_from_output(
                 new_logprobs = logprobs.slice(req_index, req_index + 1)
 
             if new_token_ids and request.use_structured_output:
-                # NOTE: structured_output_request
-                # should not be None if use_structured_output, we have
-                # check above, so safe to ignore type warning
-                request.structured_output_request.grammar.accept_tokens(  # type: ignore[union-attr]
-                    req_id, new_token_ids)
+                advance_fsm = False
+                reasoner = self.structured_output_manager.reasoner
+                is_reasoning_end_this_step = False
+
+                # NOTE: use_structured_output implies
+                # structured_output_request is not None,
+                # but type checker isn't smart enough to know this.
+                # This only affect type runtime, not actual runtime.
+                # assert is also not recommended on perf-sensitive runtime path.
+                if TYPE_CHECKING:
+                    assert request.structured_output_request is not None
+                    assert request.structured_output_request.grammar is not None
+
+                if reasoner is None or request.structured_output_request.reasoning_ended:  # noqa: E501
+                    advance_fsm = True
+                elif reasoner.is_reasoning_end(request.all_token_ids):
+                    request.structured_output_request.reasoning_ended = True
```

Comment:
Is this new variable meant to speed up the reasoning process?

Reply:
oh it is just a marker to determine when we want to fill in the bitmask

Reply:
maybe we can have a ReasoningState enum instead 🤔